django/middleware/locale.py

```diff
@@ -35,7 +35,12 @@ def process_response(self, request, response):
         urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
         i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
 
-        if response.status_code == 404 and not language_from_path and i18n_patterns_used:
+        if (response.status_code == 404 and
+                not language_from_path and
+                i18n_patterns_used and
+                (prefixed_default_language or language != settings.LANGUAGE_CODE)):
```

Comment:
I don't think it's required:
- If the language is the default language, Django should only prefix when `prefixed_default_language` is used.
- Otherwise, the language code would have been in the request `path_info` and `language_from_path` would evaluate to `True`.